File: src/models/transformers/qwen3/Qwen3RotaryEmbedding.cpp
#include "models/transformers/qwen3/Qwen3RotaryEmbedding.hpp"
#include "modules/Constant.hpp"

Qwen3RotaryEmbedding::Qwen3RotaryEmbedding(const Qwen3Config& config)
    : dim_(config.head_dim.value_or(config.hidden_size / config.num_attention_heads)),
      base_(config.rope_parameters.rope_theta)
{
}

std::pair<Tensor, Tensor> Qwen3RotaryEmbedding::forward(ggml_context* ctx, Tensor x, Tensor position_ids) {
    auto inv_freq = 1.0f / (Tensor::arange(0, dim_, 2) / (float)dim_).pow(base_);

    auto inv_freq_expanded = inv_freq[{Tensor::Slice::none(), Tensor::Slice::all(), Tensor::Slice::none()}].to(GGML_TYPE_F32);
    inv_freq_expanded = inv_freq_expanded.expand({position_ids.shape()[0], -1, 1});

    auto position_ids_expanded = position_ids[{Tensor::Slice::all(), Tensor::Slice::none(), Tensor::Slice::all()}].to(GGML_TYPE_F32);

    auto freqs = ggml_mul_mat(ctx, position_ids_expanded, inv_freq_expanded);
    freqs = ggml_permute(ctx, freqs, 0, 2, 1, 3);

    auto emb = Tensor::cat({freqs, freqs}, -1);

    return {cos(emb).to(x.dtype()), sin(emb).to(x.dtype())};
}

File: test/src/transformers/models/qwen3/Qwen3RotaryEmbedding.py
from utils import TestCase
import torch
from transformers import Qwen3Config
from transformers.models.qwen3.modeling_qwen3 import Qwen3RotaryEmbedding
from transformers.models.qwen3.modeling_qwen3 import apply_rotary_pos_emb

class TestTransformersQwen3RotaryEmbedding(TestCase):
    def test_default(self):
        config = Qwen3Config(
            hidden_size=8,
            head_dim=4,
            num_attention_heads=2,
            max_position_embeddings=128,
            rope_theta=10000.0,
        )

        pos_embed = Qwen3RotaryEmbedding(config)

        position_ids = torch.tensor([
            [0, 1, 2, 3],
        ])

        q = torch.randn(
            1,  # batch
            2,  # heads
            position_ids.shape[1],  # sequence
            config.hidden_size // config.num_attention_heads,  # head_dim
        )

        k = torch.randn_like(q)

        cos, sin = pos_embed(q, position_ids)

        expected_q, expected_k = apply_rotary_pos_emb(
            q,
            k,
            cos,
            sin,
            unsqueeze_dim=1,
        )

        actual_q = self.cli(
            "Qwen3RotaryEmbedding",
            "--head_dim", "4",
            "--rope_theta", "10000",
            "--x", str(q.tolist()),
            "--position_ids", str(position_ids.tolist()),
        )

        actual_k = self.cli(
            "Qwen3RotaryEmbedding",
            "--head_dim", "4",
            "--rope_theta", "10000",
            "--x", str(k.tolist()),
            "--position_ids", str(position_ids.tolist()),
        )

        self.assertTensors(actual_q, [expected_q])
        self.assertTensors(actual_k, [expected_k])
